securesystemslib/metadata.py

```diff
@@ -197,3 +197,28 @@ def deserialize_payload(
 
         payload = deserializer.deserialize(self.payload, class_type)
         return payload
+
+    def match_payload(
+        self,
+        class_type: Any,
+        deserializer: Optional[BaseDeserializer] = None,
+    ) -> bool:
+        """Checks for the payload if deserializes into the provided class type.
+
+        Arguments:
+            class_type: The class to be deserialized. If the default
+                deserializer is used, it must implement ``JSONSerializable``.
+            deserializer: ``BaseDeserializer`` implementation to use.
+                Default is JSONDeserializer.
+
+        Returns:
+            Boolean. Indicating the provided class type is matched with
+            payload.
+        """
+
+        try:
+            self.deserialize_payload(class_type, deserializer)
+        except exceptions.DeserializationError:
+            return False
+        else:
+            return True
```

Comment:
Why not look at `self.payload_type` to know which class to deserialize into?

Reply:
ITE-5 suggests that in-toto will have metadata with `payload_type=application/vnd.in-toto+json` and in-toto has payloads that can further classified into two types, `Link` and `Layout`.

Also, This method is common with in-toto `Metablock` which has only two types either link or layout.

Reply:
Oh, right, thanks for the reminder. Still, it seems unnecessary to fully deserialize the payload twice, given that we know exactly what we expect.

Reply:
> Oh, right, thanks for the reminder. Still, it seems unnecessary to fully deserialize the payload twice, given that we know exactly what we expect.

I am open to other solution to this problem. There is only one case where payload was deserialized twice in a row.

Reply:
> There is only one case where payload was deserialized twice in a row.

We still deserialize the full object to extract a single field, and then throw away the object even if we still need it.

> I am open to other solution to this problem.

I'd like to use a similar pattern as in `Metablock.from_dict`. It also doesn't need to be passed a `Layout` or `Link` class in order to deserialize the passed data into its `signed` field, because it knows that the passed data must have a `_type` field, whose value must be either `link` or `layout`, so it can during deserialization just call `Link.from_dict` or `Layout.from_dict`.

And the exact same thing applies to an `Envelope` when in the `in-toto` context. The problem is that `Envelope` does not only exist in the in-toto context. So we need to find a way to register an in-toto specific payload parsing for a generic `Envelope` class. I have to think about how to best do this.